test/Dialect/FIRRTL/imdce.mlir

```diff
@@ -428,3 +428,31 @@ firrtl.circuit "Top" {
     firrtl.instance baz2 sym @baz2 @Baz()
   }
 }
+
+// -----
+
+// This tests that dead modules do not keep dead instance output ports alive.
+
+firrtl.circuit "Test" {
+  firrtl.extmodule @ExtModule(out out : !firrtl.uint<1>)
+
+  // %out is only used by a dead module. It should be removed from the port list.
+  // CHECK: firrtl.module private @Blah() {
+  firrtl.module private @Blah(out %out : !firrtl.uint<1>) {
+    %extmodule_out = firrtl.instance extmodule @ExtModule(out out : !firrtl.uint<1>)
+    firrtl.strictconnect %out, %extmodule_out : !firrtl.uint<1>
+  }
+  firrtl.module @Test() attributes {convention = #firrtl<convention scalarized>} {
+    // CHECK: firrtl.instance blah interesting_name @Blah()
+    %blah_out = firrtl.instance blah interesting_name @Blah(out out : !firrtl.uint<1>)
+  }
+  // This module is dead.  Since it is the only user of @Blah's output port, the
+  // output should be deleted.
+  firrtl.module private @Other(out %out : !firrtl.uint<1>) {
+    // CHECK: %0 = builtin.unrealized_conversion_cast to !firrtl.uint<1>
```

Comment:
Ah, I didn't imagine there was a corner case where unrealized_conversin cast outlives IMDCE pass. Probably we have to just erase modules unreachable from toplevel.

Reply:
Yeah.  The FIRRTL Inliner will delete unreachable modules before this pass runs, so this isn't reachable in the firtool pipeline.  Arguably the inliner should not be deleting modules at all, but at the time it was trying to maintain the invariant that all modules are reachable from the top level, and so it was supposed to be deleting the modules where every instance was inlined.

Reply:
> at the time it was trying to maintain the invariant that all modules are reachable from the top level, and so it was supposed to be deleting the modules where every instance was inlined.

That's totally reasonable and I think that's good thing to delete such modules in the early pipeline 👍 #5517 should clean up unreachable modules in IMDCE as well.